crates/red_knot_python_semantic/src/types.rs

```diff
@@ -521,6 +521,56 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the boolean value of a type.
+    /// This is used to determine the value that would be returned when `__bool__` is called on an object
+    pub fn bool(&self, db: &'db dyn Db) -> Option<bool> {
```

Comment:
I was wondering about whether this should return a custom `Truthiness` enum, e.g.

```rs
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Truthiness {
    AlwaysTrue,
    AlwaysFalse,
    Ambiguous,
}

impl Truthiness {
    fn negate(self) -> Self {
        match self {
            Self::AlwaysTrue => Self::AlwaysFalse,
            Self::AlwaysFalse => Self::AlwaysTrue,
            Self::Ambiguous => Self::Ambiguous
        }
    }

    fn into_type(self, db: &dyn Db) -> Type {
        match self {
            Self::AlwaysTrue => Type::BooleanLiteral(true),
            Self::AlwaysFalse => Type::BooleanLiteral(false),
            Self::Ambiguous => builtins_symbol_ty(db, "bool").to_instance(db),
        }
    }
}

impl From<bool> for Truthiness {
    fn from(value: bool) -> Self {
        if value {
            Truthiness::AlwaysTrue
        } else {
            Truthiness::AlwaysFalse
        }
    }
}
```

This has a few advantages over returning `Option<bool>`:
- It's more expressive of the fact that there are really three distinct states being expressed by the return type (and expresses what these states are)
- The `negate` method will make it easier to implement support for the `not` operator in a generalised way in the future
- Returning a custom enum will make it more adaptable in the future for when we implement support in this method for arbitrary `__bool__` and `__len__` methods. Once we add that support, we may need to add some way of bubbling up errors from `Type::call()` in this method.

Reply:
Bikesheddy comment, but my preferred name for the method would also be `boolean_eval`, since we're evaluating the booleanness of the type :)

Reply:
I think the custom enum makes sense here, because it expresses the semantics clearly and allows the caller to decide whether the actual BooleanLiteral or Instance(bool) type needs to be reified or not. It fits in well with our existing pattern with `IterationOutcome` and `CallOutcome`, where these methods can can return an enum that lazily allows the caller to make decisions about how to handle the results.

I'm not such a big fan of `boolean_eval` as the name, mostly because like 90% of what we do in the type checker could be described using the word "eval" if we chose to; it's a very generic term. I think either just `bool` or `boolean` is fine.

Reply:
I'm okay with leaving it as `bool`